llama_index/vector_stores/postgres.py

```diff
@@ -0,0 +1,158 @@
+        id = Column(BIGINT, primary_key=True, autoincrement=True)
+        text = Column(VARCHAR, nullable=False)
+        extra_info = Column(JSON)
+        doc_id = Column(VARCHAR)
+        embedding = Column(Vector(1536))  # type: ignore
+
+    tablename = "data_%s" % index_name  # dynamic table name
+    class_name = "Data%s" % index_name  # dynamic class name
+    model = type(class_name, (AbstractData,), {"__tablename__": tablename})
+    return model
+
+
+class PGVectorStore(VectorStore):
+    stores_text = True
+
+    def __init__(self, connection_string: str, table_name: str) -> None:
+        try:
+            import sqlalchemy  # noqa: F401
+        except ImportError:
+            raise ImportError("`sqlalchemy` package not found")
+
+        try:
+            import pgvector  # noqa: F401
+        except ImportError:
+            raise ImportError("`pgvector` package not found")
+
+        try:
+            import psycopg2  # noqa: F401
+        except ImportError:
+            raise ImportError("`psycopg2-binary` package not found")
+
+        from sqlalchemy.orm import declarative_base
+
+        self.connection_string = connection_string
+        self.table_name: str = table_name.lower()
+        self._conn = self._connect()
+        self._base = declarative_base()
+        # sqlalchemy model
+        self.table_class = get_data_model(self._base, self.table_name)
+        self._create_extension()
+        self._create_tables_if_not_exists()
+
+    @classmethod
+    def conn_str_from_params(
+        cls, host: str, port: int, database: str, user: str, password: str
+    ) -> str:
+        """Return connection string from database parameters."""
+        return f"postgresql+psycopg2://{user}:{password}@{host}:{port}/{database}"
+
+    def _connect(self) -> Any:
+        import sqlalchemy
+
+        engine = sqlalchemy.create_engine(self.connection_string)
+        conn: sqlalchemy.engine.Connection = engine.connect()
+        return conn
+
+    def _create_tables_if_not_exists(self) -> None:
+        with self._conn.begin():
+            self._base.metadata.create_all(self._conn)
+
+    def _create_extension(self) -> None:
+        from sqlalchemy.orm import Session
+        import sqlalchemy
+
+        with Session(self._conn) as session:
+            statement = sqlalchemy.text("CREATE EXTENSION IF NOT EXISTS vector")
+            session.execute(statement)
+            session.commit()
+
+    def add(self, embedding_results: List[NodeWithEmbedding]) -> List[str]:
+        from sqlalchemy.orm import Session
+
+        ids = []
+        with Session(self._conn) as session:
+            for result in embedding_results:
+                ids.append(result.id)
+
+                item = self.table_class(
+                    doc_id=result.id,
+                    embedding=result.embedding,
+                    text=result.node.text,
+                    extra_info=result.node.extra_info,
```

Comment:
Could you use from `node_to_metadata_dict` in `llama_index/vector_stores/utils.py` to construct the metadata dict?

You can see `llama_index/vector_stores/chroma.py` for an example

Reply:
Done